---
output: html_document
params:
  date: !r Sys.Date()
  set_title: "Fairfax_2012 Micro Array study QC report"
  set_author: "Nurlan Kerimov"
  quant_method: "HumanHT-12_V4"
  sample_meta_path: "/home/kerimov/work/SampleArcheology/studies/cleaned/Fairfax_2012.tsv"
  count_matrix_path: "/home/kerimov/work/feature_counts_QC/data/counts/micro_array/raw/Fairfax_2012.tsv.gz"
  projections: "/home/kerimov/work/feature_counts_QC/data/popAssign/results_Fairfax_2014/new_dataset_scores.profile.adj"
  pca_table: "/home/kerimov/work/feature_counts_QC/data/popAssign/results_Fairfax_2014/main_overlapped_pca.vect"
  phenotype_meta_path: "/home/kerimov/work/feature_counts_QC/data/annotations/HumanHT-12_V4_Ensembl_96_phenotype_metadata.tsv.gz"
  source_populations_file: "/home/kerimov/work/feature_counts_QC/data/popAssign/igsr_samples.tsv"
  eqtl_utils_path: "/home/kerimov/work/eQTLUtils"
  use_plotly: FALSE
---

---
title: `r params$set_title`
author: `r params$set_author`
---

- [Plot PCA Analysis](#plot-pca-analysis)
- [Plot the MDS analysis](#plot-the-mds-analysis)
- [Sex-specific gene expression analysis](#sex-specific-gene-expression-analysis)
- [popAssign Genotype PCA analysis](#popassign-genotype-pca-analysis)

***
```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.height = 6, fig.width = 10, fig.align = "center")
```

```{r load_libraries_and_inputs, echo=FALSE, results="hide", include=FALSE}
suppressPackageStartupMessages(library("readr"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("devtools"))
suppressPackageStartupMessages(library("tidyr"))
suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("SummarizedExperiment"))
suppressPackageStartupMessages(library("data.table"))
suppressPackageStartupMessages(library("plotly"))
suppressPackageStartupMessages(library("knitr"))

sample_meta_path <- params$sample_meta_path
count_matrix_path <- params$count_matrix_path
projections <- params$projections
pca_table <- params$pca_table
use_plotly <- params$use_plotly

phenotype_meta_path <- params$phenotype_meta_path
source_populations_file <- params$source_populations_file
devtools::load_all(params$eqtl_utils_path)
```

```{r debug_chunk, echo=FALSE, results="hide", include=FALSE, eval=FALSE}
set_title = "Fairfax_2012 MicroArray QC report"
quant_method = "HumanHT-12_V4"
eqtl_utils_path = "/home/kerimov/work/eQTLUtils"
sample_meta_path = "/home/kerimov/work/SampleArcheology/studies/cleaned/Fairfax_2012.tsv"
count_matrix_path = "/home/kerimov/work/feature_counts_QC/data/counts/micro_array/raw/Fairfax_2012.tsv.gz"
projections = "/home/kerimov/work/feature_counts_QC/data/popAssign/results_Fairfax_2014/new_dataset_scores.profile.adj"
pca_table = "/home/kerimov/work/feature_counts_QC/data/popAssign/results_Fairfax_2014/main_overlapped_pca.vect"
phenotype_meta_path = "/home/kerimov/work/feature_counts_QC/data/annotations/HumanHT-12_V4_Ensembl_96_phenotype_metadata.tsv.gz"
source_populations_file = "/home/kerimov/work/feature_counts_QC/data/popAssign/igsr_samples.tsv"
```


```{r read_data, echo=FALSE, warning=FALSE}
# Read the inputs
sample_metadata <- utils::read.csv(sample_meta_path, sep = '\t', stringsAsFactors = FALSE)
phenotype_meta = readr::read_delim(phenotype_meta_path, delim = "\t", col_types = "ccccciiicciidi")
data_fc <- utils::read.csv(count_matrix_path, sep = '\t')

dummy <- assertthat::assert_that(params$quant_method %in% c("HumanHT-12_V4", "quant_method"), msg = "Quantification method is not a valid! Please choose 'HumanHT-12_V4' or 'gene_counts'")

se <- eQTLUtils::makeSummarizedExperimentFromCountMatrix(assay = data_fc, 
                                                         row_data = phenotype_meta, 
                                                         col_data = sample_metadata, 
                                                         quant_method = params$quant_method)

message("Number of samples in study before Quality Control: ", sample_metadata %>% nrow())
message("Number of samples in study after Quality Control: ", sample_metadata[sample_metadata$rna_qc_passed & sample_metadata$genotype_qc_passed,] %>% nrow())
```

### Plot PCA Analysis

Calculate the PCA matrix and plot the PCA analysis output

* Quantile Normalised

* Phenotypes having median less than 1 are filtered out

* log2(matrix + 0.1) is applied

```{r pca_res, echo=FALSE}
valid_gene_types = c("lincRNA","protein_coding","IG_C_gene","IG_D_gene","IG_J_gene",
                       "IG_V_gene", "TR_C_gene","TR_D_gene","TR_J_gene", "TR_V_gene",
                       "3prime_overlapping_ncrna","known_ncrna", "processed_transcript",
                       "antisense","sense_intronic","sense_overlapping")
valid_chromosomes = c("1","10","11","12","13","14","15","16","17","18","19",
                      "2","20","21","22","3","4","5","6","7","8","9")

# choose only valid gene types, valid chromosomes and TPM normalise
processed_se = eQTLUtils::filterSummarizedExperiment(se,
                                           valid_chromosomes = valid_chromosomes,
                                           valid_gene_types = valid_gene_types, 
                                           filter_rna_qc = TRUE, 
                                           filter_genotype_qc = TRUE) 


processed_se = processed_se %>% eQTLUtils::array_normaliseSE(adjust_batch = TRUE)  
processed_se = processed_se[apply(SummarizedExperiment::assays(processed_se)$norm_exprs, 1, median) > 1, ]
message("Number of phenotypes: ", processed_se %>% nrow())

study_name <- processed_se$study[1]
```


```{r transform_se, echo=FALSE}
n_pcs = 10
if (ncol(processed_se) < 10)
  { n_pcs = ncol(processed_se) }

#Perform PCA
pca_res = eQTLUtils::transformSE_PCA(processed_se, assay_name = "norm_exprs", n_pcs = n_pcs, log_transform = TRUE, center = TRUE, scale. = TRUE)
plot_title_text <- paste0(study_name, " PCA - quantile normalized, log2 | Sample Size: ", nrow(processed_se %>% SummarizedExperiment::colData()))

PCA.plot <- ggplot2::ggplot(pca_res$pca_matrix, ggplot2::aes(x = PC1, y = PC2, color = qtl_group, label = sample_id)) +
    ggplot2::geom_point() +
    ggplot2::scale_shape_manual(values=c(20,17,18,11,14,25,8,seq(0,7))) +
    ggplot2::labs(x=paste0("PC 1 - (", round(pca_res$var_exp[1]*100, digits = 1),"% var. explained)"),
                  y=paste0("PC 2 - (", round(pca_res$var_exp[2]*100, digits = 1),"% var. explained)"),
                  title = plot_title_text) +
    ggplot2::theme(plot.title =  ggplot2::element_text(size = 12, face="italic")) +
    ggplot2::theme_bw()

if (use_plotly) {
  plotly::ggplotly(PCA.plot, width=1000)
} else {
  PCA.plot
}
```

### Plot the MDS analysis

* Quantile Normalised

* Phenotypes having median less than 1 are filtered out

* log2(matrix + 0.1) is applied

* dist = cor(matrix, method = "pearson")

* isoMDS(1-dist, k=2) is applied as MDS distance, k=2 scaled into 2 dimensions 

```{r mds, results="hide", warning=FALSE, echo=FALSE, collapse=TRUE}
#Perform MDS

message("Number of phenotypes after filtering out phenotypes with median value less than one [median < 1]: ", processed_se %>% nrow())

matrix = log(SummarizedExperiment::assays(processed_se)$norm_exprs+0.1,2)  
plot_title_text <- paste0(study_name, " MDS - quantile normalized, log2 | Sample Size: ", nrow(processed_se %>% SummarizedExperiment::colData()))

dist = cor(matrix, method = "pearson")
fit <- MASS::isoMDS(1-dist, k=2)

mds_matrix = SummarizedExperiment::as.data.frame(fit$points) %>%
  as_tibble() %>% dplyr::mutate(sample_id = rownames(fit$points)) %>%
  dplyr::left_join(SummarizedExperiment::as.data.frame(SummarizedExperiment::colData(processed_se)), by = "sample_id")

mds_plot = ggplot2::ggplot(mds_matrix, ggplot2::aes(x = V1, y = V2, color = qtl_group, label = sample_id)) +
    ggplot2::geom_point() +
    ggplot2::scale_shape_manual(values=c(20,17,18,11,14,25,8,seq(0,7))) +
    ggplot2::labs(x="MDS Coordinate 1", y="MDS Coordinate 2", title = plot_title_text) +
    ggplot2::theme(plot.title = element_text(size = 12, face="italic")) +
    ggplot2::theme_bw()
```

```{r mds_plot, echo=FALSE}
if (use_plotly) {
  plotly::ggplotly(mds_plot, width=1000)
} else {
  mds_plot
}
```

### Sex-specific gene expression analysis

* Quantile Normalised

* log2(matrix + 1) is applied

```{r sex_specific, results="hide", warning=FALSE, echo=FALSE, collapse=TRUE, include=FALSE}
se <- eQTLUtils::filterSummarizedExperiment(se, filter_rna_qc = TRUE, filter_genotype_qc = TRUE)
sex_qc_df <- eQTLUtils::array_CalculateSexQCDataFrame(se, assay_name = "exprs",
                                                          filterByCondition = FALSE,
                                                          normalise_array = TRUE)

if (all(is.na(sex_qc_df$sex))) { sex_qc_df$sex <- "NA"}

Sex_QC_plot <- ggplot2::ggplot(sex_qc_df, ggplot2::aes(x=XIST, y=Y_chrom_mean, label = sample_id)) +
  ggplot2::geom_point(ggplot2::aes(col=sex)) +
  ggplot2::labs(x="Expression XIST", y="Mean Expression genes on Y", 
                title = paste0(study_name, " DS | quantile normalised, log2, | Sample Size: ", nrow(sex_qc_df))) +
  ggplot2::theme_bw()
```

```{r plot_sex_qc, echo=FALSE}
if (use_plotly) {
  plotly::ggplotly(Sex_QC_plot, width=1000)
} else {
  Sex_QC_plot
}
```


### popAssign Genotype PCA analysis
```{r popAssign, echo=FALSE}
pca <- read.table(pca_table, header = FALSE, )
pca <- pca %>% select(-c(2)) %>% rename(genotype_id=V1, PC1=V3, PC2=V4, PC3=V5)

source_populations <- read.table(source_populations_file, header = TRUE, sep='\t')
main_pca <- merge(pca, select(source_populations, genotype_id, superpopulation_code), by  = "genotype_id") 

projections_pcs <- read.table(projections, header = TRUE)
projections_pcs <- projections_pcs %>% select(-c(2,3,4)) %>% 
  rename(genotype_id=ID1, PC1=Adjusted1, PC2=Adjusted2, PC3=Adjusted3) %>% 
  mutate(superpopulation_code = study_name)

comb_pcs = rbind(main_pca, projections_pcs)
projected_plot <- ggplot(comb_pcs, aes(x=PC1, y=PC2, color=superpopulation_code)) + geom_point() + 
  guides(color = guide_legend(title='Superpopulation')) + ggtitle(study_name) +
  ggplot2::theme_bw()

message("Number of genotypes in study ", study_name, ": ", projections_pcs %>% nrow())
message("Number of genotypes in reference data: ", main_pca %>% nrow())

if (use_plotly) {
  plotly::ggplotly(projected_plot)
} else {
  projected_plot
}
```
